chainfactory.go

```diff
@@ -108,107 +53,89 @@ var builtinChainConfigs = map[string]ibc.ChainConfig{
+func NewBuiltinChainFactory(log *zap.Logger, specs []*ChainSpec) *BuiltinChainFactory {
+	return &BuiltinChainFactory{log: log, specs: specs}
 }
 
 func (f *BuiltinChainFactory) Count() int {
-	return len(f.entries)
+	return len(f.specs)
 }
 
 func (f *BuiltinChainFactory) Chains(testName string) ([]ibc.Chain, error) {
-	chains := make([]ibc.Chain, len(f.entries))
-	for i, e := range f.entries {
-		chain, err := e.GetChain(f.log, testName)
+	chains := make([]ibc.Chain, len(f.specs))
+	for i, s := range f.specs {
+		cfg, err := s.Config()
+		if err != nil {
+			// Prefer to wrap the error with the chain name if possible.
+			if s.Name != "" {
+				return nil, fmt.Errorf("failed to build chain config %s: %w", s.Name, err)
+			}
+
+			return nil, fmt.Errorf("failed to build chain config at index %d: %w", i, err)
+		}
+
+		chain, err := buildChain(f.log, testName, *cfg, s.NumValidators, s.NumFullNodes)
 		if err != nil {
 			return nil, err
 		}
 		chains[i] = chain
 	}
+
 	return chains, nil
 }
 
-func (f *BuiltinChainFactory) Name() string {
-	parts := make([]string, len(f.entries))
-	for i, e := range f.entries {
-		parts[i] = e.Name + "@" + e.Version
-	}
-	return strings.Join(parts, "+")
-}
+const (
+	defaultNumValidators = 2
+	defaultNumFullNodes  = 1
+)
 
-func (f *BuiltinChainFactory) Labels() []label.Chain {
-	labels := make([]label.Chain, len(f.entries))
-	for i, e := range f.entries {
-		label := label.Chain(e.Name)
-		if !label.IsKnown() {
-			// The label must be known (i.e. registered),
-			// otherwise filtering from the command line will be broken.
-			panic(fmt.Errorf("chain name %s is not a known label", e.Name))
-		}
-		labels[i] = label
+func buildChain(log *zap.Logger, testName string, cfg ibc.ChainConfig, numValidators, numFullNodes *int) (ibc.Chain, error) {
+	nv := defaultNumValidators
+	if numValidators != nil {
+		nv = *numValidators
+	}
+	nf := defaultNumFullNodes
+	if numFullNodes != nil {
+		nf = *numFullNodes
 	}
-	return labels
-}
-
-// CustomChainFactory is a ChainFactory that supports returning chains that are defined by ChainConfig values.
-type CustomChainFactory struct {
-	entries []CustomChainFactoryEntry
-	log     *zap.Logger
-}
-
-// CustomChainFactoryEntry describes a chain to be returned by a CustomChainFactory.
-type CustomChainFactoryEntry struct {
-	Config        ibc.ChainConfig
-	NumValidators int
-	NumFullNodes  int
-}
-
-// NewCustomChainFactory returns a CustomChainFactory that returns chains defined by entries.
-func NewCustomChainFactory(entries []CustomChainFactoryEntry, logger *zap.Logger) *CustomChainFactory {
-	return &CustomChainFactory{entries: entries, log: logger}
-}
 
-func (e CustomChainFactoryEntry) GetChain(testName string, log *zap.Logger) (ibc.Chain, error) {
-	switch e.Config.Type {
+	switch cfg.Type {
 	case "cosmos":
-		return cosmos.NewCosmosChain(testName, e.Config, e.NumValidators, e.NumFullNodes, log), nil
+		return cosmos.NewCosmosChain(testName, cfg, nv, nf, log), nil
```

Comment:
I keep thinking how to remove the import cycle caused by this (and potentially `penumbra` package) line. I'd love the helpers in the `test` package to be at root `ibctest`. But importing `cosmos` in `ibctest` causes the cycle because cosmos uses the helpers like `WaitForBlocks`